File: src/ark_recognizer_py/recognize_type.py
import math
from dataclasses import dataclass

import cv2 as cv
import numpy as np
from numpy.typing import NDArray

from ark_recognizer_py.item import ITEMS
from ark_recognizer_py.item import Item
from ark_recognizer_py.recognize_icon import IRResult
from ark_recognizer_py.utils import clamp
from ark_recognizer_py.utils import img_size


@dataclass
class TRResult:
    """
    Represents a type recognition result.
    """
    item: Item
    size: NDArray
    loc: NDArray
    val: float


TM_SCALE = 1 / 3
TM_THRESHOLD = 0.6
TEMPL_ICON_SIDE_LENGTH = 183
TEMPL_CIRCLE_RADIUS = 163 / 2


def __preprocess_items_template():
    items_template = {}
    for item in ITEMS:
        template = item.item_icon
        template = cv.cvtColor(template, cv.COLOR_BGRA2BGR)
        template = cv.resize(template, None,
                             fx=TM_SCALE,
                             fy=TM_SCALE,
                             interpolation=cv.INTER_CUBIC)
        items_template[item.item_id] = template
    return items_template


def __preprocess_items_templates_mask():
    items_template_mask = {}
    for item in ITEMS:
        mask = item.item_icon[:, :, 3]
        _, mask = cv.threshold(mask, 170, 255, cv.THRESH_BINARY)
        mask = cv.resize(mask, None,
                         fx=TM_SCALE,
                         fy=TM_SCALE,
                         interpolation=cv.INTER_CUBIC)
        items_template_mask[item.item_id] = mask
    return items_template_mask


ITEMS_TEMPLATE: dict[str, NDArray] = __preprocess_items_template()
ITEMS_TEMPLATE_MASK: dict[str, NDArray] = __preprocess_items_templates_mask()


def match_item_template(icon_image: NDArray, item: Item) -> TRResult:
    tmpl = ITEMS_TEMPLATE[item.item_id]
    tmpl_mask = ITEMS_TEMPLATE_MASK[item.item_id]

    icon_w, icon_h = img_size(icon_image)
    tmpl_w, tmpl_h = img_size(tmpl)
    if icon_h < tmpl_h or icon_w < tmpl_w:
        return TRResult(item, np.zeros(0), np.zeros(0), -math.inf)

    result: NDArray = cv.matchTemplate(icon_image,
                                       templ=tmpl,
                                       mask=tmpl_mask,
                                       method=cv.TM_CCOEFF_NORMED)
    _, max_val, _, max_loc = cv.minMaxLoc(result)
    tr_result = TRResult(item, img_size(tmpl), np.array(max_loc), max_val)
    return tr_result


def match_all_item_templates(icon_image: NDArray) -> TRResult:
    # match icon image with all item templates
    results = []
    for item in ITEMS:
        result = match_item_template(icon_image, item)
        results.append(result)

    result = max(results, key=lambda x: x.val)
    return result


def recognize_type(scene_image: NDArray, circles: list[IRResult]) -> list[TRResult]:
    results = []
    for circle in circles:
        # the x, y, and radius of this icon circle
        circle_x, circle_y, circle_radius = np.rint(circle).astype(int)
        # the side length of this icon. for tolerant, * 1.1
        icon_side_length = circle_radius * (TEMPL_ICON_SIDE_LENGTH / TEMPL_CIRCLE_RADIUS) * 1.1

        w, h = img_size(scene_image)
        x0 = clamp(round(circle_x - icon_side_length / 2), 0, w)
        x1 = clamp(round(circle_x + icon_side_length / 2), 0, w)
        y0 = clamp(round(circle_y - icon_side_length / 2), 0, h)
        y1 = clamp(round(circle_y + icon_side_length / 2), 0, h)
        # the icon image cropped from the scene image
        icon_image = scene_image[y0:y1, x0:x1]

        # check if the icon cropped is empty
        if len(icon_image) == 0:
            continue

        # the scale factor to scale this icon size to template icon size
        icon_scale_factor = (TEMPL_CIRCLE_RADIUS / circle_radius) * TM_SCALE
        # the icon image cropped and resized from scene image
        icon_image = cv.resize(icon_image, None, fx=icon_scale_factor, fy=icon_scale_factor)

        result = match_all_item_templates(icon_image)
        if result.val < TM_THRESHOLD:
            continue

        # calculate the actual size and location in the scene image
        result.size = np.rint(result.size / icon_scale_factor).astype(int)
        result.loc = np.rint((x0, y0) + result.loc / icon_scale_factor).astype(int)

        results.append(result)

    return results

File: src/ark_recognizer_cpp/recognize_type.cpp
//
// Created by Flandia on 2021/12/12.
//

#include <opencv2/imgproc.hpp>
#include "recognize_type.h"

using namespace std;
using namespace cv;

std::map<std::string, cv::Mat> preprocess_item_templates(const std::vector<Item> &items) {
    map<string, Mat> templs;
    for (const auto &item: items) {
        auto templ = item.item_icon.clone();
        cvtColor(templ, templ, COLOR_BGRA2BGR);
        resize(templ, templ, Size(), TM_SCALE, TM_SCALE, INTER_CUBIC);
        templs[item.item_id] = templ;
    }
    return templs;
}

std::map<std::string, cv::Mat> preprocess_item_template_masks(const std::vector<Item> &items) {
    map<string, Mat> masks;
    for (const auto &item: items) {
        vector<cv::Mat> channels;
        cv::split(item.item_icon.clone(), channels);

        auto mask = channels[3];
        threshold(mask, mask, 170, 255, THRESH_BINARY);
        resize(mask, mask, Size(), TM_SCALE, TM_SCALE, INTER_CUBIC);
        masks[item.item_id] = mask;
    }
    return masks;
}

TRResult match_item_template(const Mat &icon_image,
                             const Item &item,
                             const Mat &templ, const Mat &mask) {
    auto icon_w = icon_image.cols, icon_h = icon_image.rows;
    auto templ_w = templ.cols, templ_h = templ.rows;
    if (icon_w < templ_w || icon_h < templ_h) return TRResult{item, Point(), Size(), -HUGE_VAL};

    Mat result;
    matchTemplate(icon_image, templ, result, TM_CCOEFF_NORMED, mask);
    double max_val;
    Point max_loc;
    minMaxLoc(result, nullptr, &max_val, nullptr, &max_loc);

    auto tr_result = TRResult{item, max_loc, templ.size(), max_val};
    return tr_result;
}

TRResult match_all_item_templates(const Mat &icon_image,
                                  const vector<Item> &items,
                                  map<string, Mat> templs,
                                  map<string, Mat> masks) {
    vector<TRResult> results;
    for (const auto &item: items) {
        auto templ = templs[item.item_id];
        auto mask = masks[item.item_id];
        auto result = match_item_template(icon_image, item, templ, mask);
        results.push_back(result);
    }

    auto result = *max_element(
            results.begin(), results.end(),
            [&](TRResult &tr1, TRResult &tr2) -> bool { return tr1.val < tr2.val; }
    );
    return result;
}

std::vector<TRResult> recognize_type(const cv::Mat &scene_image, const std::vector<IRResult> &ir_results,
                                     const std::vector<Item> &items,
                                     const std::map<std::string, cv::Mat> &templates,
                                     const std::map<std::string, cv::Mat> &masks) {
    vector<TRResult> results;
    for (const auto &ir_result: ir_results) {
        double circle_x = ir_result.x, circle_y = ir_result.y, circle_radius = ir_result.radius;
        double icon_side_length = circle_radius * (TEMPL_ICON_SIDE_LENGTH / TEMPL_CIRCLE_RADIUS) * 1.1;

        int w = scene_image.cols, h = scene_image.rows;
        int x0 = clamp((int) round(circle_x - icon_side_length / 2), 0, w);
        int x1 = clamp((int) round(circle_x + icon_side_length / 2), 0, w);
        int y0 = clamp((int) round(circle_y - icon_side_length / 2), 0, h);
        int y1 = clamp((int) round(circle_y + icon_side_length / 2), 0, h);

        auto icon_image = scene_image(Range(y0, y1), Range(x0, x1));
        if (icon_image.empty()) continue;

        auto icon_scale_factor = (TEMPL_CIRCLE_RADIUS / circle_radius) * TM_SCALE;
        resize(icon_image, icon_image, Size(), icon_scale_factor, icon_scale_factor);

        auto result = match_all_item_templates(icon_image, items, templates, masks);
        if (result.val < TM_THRESHOLD) continue;

        result.loc = Point((int) round(x0 + result.loc.x / icon_scale_factor),
                           (int) round(y0 + result.loc.y / icon_scale_factor));
        result.size = Size((int) round(result.size.width / icon_scale_factor),
                           (int) round(result.size.height / icon_scale_factor));

        results.push_back(result);
    }
    return results;
}
